### DISCUSSION TITLE
[archived] "Container Set Template" docs discussion

### DISCUSSION THREAD

**Discussion by giscus** [argoproj/argo-workflows#10552] [General]
# argo-workflows/container-set-template/

https://argoproj.github.io/argo-workflows/container-set-template/

**Comment by satya-knskp** [comment]
Hi Can we use suspend step in conatiner steps?

**Reply by satya-knskp** [reply]
```yaml
- name: approve
  suspend: {}
```

https://argoproj.github.io/argo-workflows/walk-through/suspending/

**Reply by agilgur5** [reply]
Yea that ^link is the correct docs reference. `suspend` is a different Template Type entirely from `containerSet` -- for clarity, I've added it to the list of template types as well: #11500

**Comment by freegroup** [comment]
unfortunately, the first pods of the container workflow demo ends with an error like
```

Error (exit code 1): pods "container-set-template-7prl5" is forbidden: User "system:serviceaccount:default:default" cannot patch resource "pods" in API group "" in the namespace "default"

```

should I can change the serviceAccount for the workflow to "argo"?

**Reply by agilgur5** [reply]
It depends on your set-up. See https://argo-workflows.readthedocs.io/en/latest/service-accounts/ and https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/

**Comment by nicolas-vivot** [comment]
This documentation about Resources is so wrong...

Guys, if you only specify requests on resources, you will ends up each of your container potentially using all CPU of the node your Pod/container is running.

Giving only resource request do not allow you to control the maximum of that resources your process will use at a time.

If you want to control this, you must specify both request and limit on resources. And by doing so, what will happen is the following:
- container 1 has 2 cpu request and 2 cpu limit
- container 2 has 2 cpu request and 2 cpu limit

You Pod containing your containerSet will then request 4 CPU in total, and:
- if your container runs in //, they will each use at most 2 CPU
- if your container runs sequentially, each will use at most 2 CPU (not 4 as this documentation specify). In this case, you will always consume half the total resources allocated to your Pod.

From this, you can easily image what happen if you have 8 containers running in sequence in a container set, with each 2 CPU: you will book and pay for 16 CPU but only consume 2 CPU at any time.
Same goes with any resource type.

Please update your documentation not to confuse users or lead them into a situation were they overpay.

This also means that the only scenario where containerSet really works and fit is when you have containers running in parallel and with fixed resources allocated.
